### ISSUE TITLE
Modernize python script to use currently supported imports

### ISSUE THREAD

**Issue by WSLUser** [jtesta/ssh-audit#23]
This tool is designed to increase security but itself is insecure and uses unsupported imports.

The following changes need to be made to get it at the very least python 3.6 compatible: 

Update get-opt to argparse (or choose something else like click but argparse is the easiest supported transition)

Remove future, base64 and others that are only needed below python3.6.

The following projects will likely help (including ironically the use of futurize):
https://github.com/PythonCharmers/python-future
https://github.com/asottile/pyupgrade (run all but the py37-plus flag)

You may also find https://github.com/python-poetry/poetry helpful for dependency management.

Also try to ensure the script is PEP compliant using mypy and pylint.

**Comment by jtesta** [comment]
> This tool is designed to increase security but itself is insecure and 
> uses unsupported imports.

How, exactly?


> The following changes need to be made to get it at the very least python 
> 3.6 compatible:

It seems to already work on Python 3.6:

$ python3 --version
Python 3.6.5
$ ./ssh-audit.py www.positronsecurity.com
# general
(gen) banner: SSH-2.0-OpenSSH_7.2p2 Ubuntu-4ubuntu2.8
(gen) software: OpenSSH 7.2p2
(gen) compatibility: OpenSSH 6.5+, Dropbear SSH 2013.62+
(gen) compression: enabled (zlib@openssh.com)
[...]

If you could show steps on reproducing a problem, however, please let me 
know!


> Update get-opt to argparse (or choose something else like click but 
> argparse is the easiest supported transition)

Is there any particular reason why get-opt should be replaced?  I see no 
documentation that claims it is deprecated.


> Remove future, base64 and others that are only needed below python3.6.

Python 3.5 is used by Ubuntu 16.04 LTS, which is still supported for 
another year.  And after that, it'll be supported until 2024 under paid 
contract.  Some users might appreciate Python 3.5 support left in.


> The following projects will likely help (including ironically the use of 
> futurize):
> https://github.com/PythonCharmers/python-future
> https://github.com/asottile/pyupgrade (run all but the py37-plus flag)

I suppose I'm still not clear on what problem you've run into.

I'm a bit short on time at the moment, so I can't research those 
projects.  Could you summarize what they do and/or show their sample 
output?  That would be a big help!


> You may also find https://github.com/python-poetry/poetry helpful for 
> dependency management

I'm not (yet?) seeing the problem regarding how modules are used or 
maintained.  Can you put it into more concrete terms?


-- 
Joseph S. Testa II
Founder & Principal Security Consultant
Positron Security

**Comment by WSLUser** [comment]
It's more code-health for most of this. get-opt was superseded by optparse, which itself was superseded by argparse. This will help performance for one (optparse is a decent CPU consumer but get-opt is even older. I know from experience with another python script I run). I was more concerned about CentOS 6 than Ubuntu 16.04 but CentOS 6 is getting depreciated later this year. You can install python3.6 on Ubuntu 16.04: https://askubuntu.com/questions/865554/how-do-i-install-python-3-6-using-apt-get#865569. Best recommendation as mentioned on there is to use pyenv. Poetry takes care of your pip management (including what happens in your pyenvs). The other 2 projects are best understood by reading their READMEs. At the very least, we want to drop python2 support and naming conventions. Why? Because it's officially depreciated. That means no patches for security holes or bug fixes. Also python3.5 does have some backports from 3.6 such as f-string support. We shouldn't support versions of python that the python maintainers themselves don't. Let the distro maintainers fix those issues, that's something they're required to do. Of course anybody can help contribute to fixing those distros up.

**Comment by WSLUser** [comment]
> Is there any particular reason why get-opt should be replaced?  I see no
documentation that claims it is deprecated.

https://docs.python.org/3/howto/argparse.html

**Comment by jtesta** [comment]
> It's more code-health for most of this. get-opt was superseded by 
> optparse, which itself was superseded by argparse.

I prefer to use argparse for my own projects, actually.  But I inherited 
ssh-audit as an existing project.  I don't mind making significant 
changes to help readability & maintainability, but to be honest, there 
are bigger fish to fry in this code base than migrating get-opt to argparse.

Migrating this isn't a priority of mine, but I'd be willing to merge a 
patch.


> This will help 
> performance for one (optparse is a decent CPU consumer but get-opt is 
> even older.

Unless someone is in a very special case scenario, like spawning 10K+ 
instances per second, I'm having a hard time seeing how improving 
parsing times from, say, 0.0011 seconds to 0.0009 seconds would make a 
practical difference.

As long as the difference in parse times is under 0.2 seconds, I doubt 
any users would notice or benefit.


> You can install python3.6 on Ubuntu 
> 16.04: 
> https://askubuntu.com/questions/865554/how-do-i-install-python-3-6-using-apt-get#865569.

Some people (myself included) are very wary of installing from 
strangers' PPAs.  That's a quick way to get malware'd.

And some organizations have explicit policies barring this.


> Best recommendation as mentioned on there is to use pyenv. Poetry takes 
> care of your pip management (including what happens in your pyenvs). The 
> other 2 projects are best understood by reading their READMEs.

Pro tip: volunteers--especially busy volunteers--won't always take 
initiative to do something if they don't immediately understand the 
benefits of it.  It doesn't grab their attention, so it gets pushed off 
until "later"... which may be in weeks, months, or never.

I'm not trying to be difficult; I'm just trying to show you that clear, 
succinct summaries will get you what you want faster.

If you can demonstrate a *practical* problem, with evidence that shows 
this particular path remediates it, then that would get my attention for 
sure.  But you don't necessarily need me in order to move forward with 
it.  If it addresses a concern of yours, you can submit a PR.  As long 
as it isn't counter-productive in some way, I'd probably merge it.


> At the 
> very least, we want to drop python2 support and naming conventions. Why? 
> Because it's officially depreciated. That means no patches for security 
> holes or bug fixes.

I've already informally dropped Python2 support.  I don't do any testing 
to ensure it remains working, so it can break at any time (it might 
already be broken and I wouldn't know!).

I wouldn't mind explicitly removing Python2 support, but this just 
hasn't yet fit into my priority list.  I'd accept a patch, however.


> Also python3.5 does have some backports from 3.6 
> such as f-string support.

How would this relate to the ssh-audit project?  Again, if you can break 
it down into a *practical* benefit, that would get my attention.


-- 
Joseph S. Testa II
Founder & Principal Security Consultant
Positron Security

**Comment by jtesta** [comment]
Closing this since it doesn't clearly describe a specific issue.  Feel free to open another issue that concretely illustrates code-level problems.
